### ISSUE TITLE
Ideas and Suggestions for the game

### ISSUE THREAD

**Issue by jonpchin** [UCF-Software-Architects/Sci-Fi-MUD#1]
Here is where we can discuss ideas for the Sci-FI MUD game. 
For now this will be a single player game that is run on a terminal. 
Maybe later we can make it multiplayer. These are ideas and are not set in stone.
We need ideas for a storyline.  Feel free to suggest new things. 

Classes:
     Cyborg: Cyborgs are biomechanical inhabitants of the future.  
	 They have the ability to modify their bodies with mechanical enhancements.  
		
		Pros: They have enhanced speed can use energy guns.
		
		Cons: Since their half human they have human limitations such as hunger
		thirst and fatigue.
		
		Skills: 
			-Punch
			-Kick
			-Run
			-Electronics
			-Stomp
			-Evalute
			-Self Repair
			-Projectiles
			-Energy Field
			-Turbo Boost
			-Teleport
			-Cannon Blast
			-QuickBoot
			-Motion Sensor
			-Discharge
			-Analyze
			-Reboot
			-Reconfigure
			-Hacking
			-Bioscan
			-Hyperscan
			-Cyberscan
			-Cybo Repair
			-Cybo Surgery
			-Self Destruct
			-Overdrive
		Weapons:
			- Any human weapons
			- Energy Saber
			- Energy Gun
			- Laser Gun
			- Plasma Gun
			- Laser Cannon
			- Plasma Cannon
			
	NanoMedic: NanoMedics are the future of doctors. These specialized doctors go through
	many years of training in order to master the art of building 
	nanomaterials, nanoprobes, nanoelectronic sensors and biological devices
	for the purpose of healing.
	
	Pros: Can heal many diseases and build skin-healing nanobots.
	
	Cons: Limited offensive skills/weapons.
	
	Skills:
		-Slice
		-Puncture
		-Heal
		-Diagnose
		-Tissue Enginering
		-DNA Enginering
		-Protein Engineering
		-Nanoprobe Assembly
		-Cure blind
		-Cure deaf
		-Cure disease
		-Surgery
		-Self Regeneration
		-Heart Transplant
		-Heart Regeneration
		-Organ Mobility
		-Blood Transfusion
		-NanoShield
		-NanoSensor Assembly
		
	Weapons:	
		-Scalpel
		-Surgical Saw
		-Catlin
	
	
	Cyber Seurity Architect: These experts are specialized in hacking, computer security, 
	and guarding digital networks. 
	
	Pros: Vast knowledge and skillset when working with computers. 
	Has ability to break into secure networks or and computer system in a 
	blink of an eye and can program viruses in their sleep.
	
	Cons: Physical offensive/defensive skills are severly lacking.

	Skills:
		-Software Repair
		-Hardware Repair
		-Software Enginering
		-Hardware Enginering
		-Programming Languages
		-Database Management
		-System Applications Design
		-Operating System Design
		-Script
		-Adware 
		-RootKit 
		-Trojan 
		-Worm
		-DDoS
		-Antivirus Scan
		-Firewall
		-Proxy
		-IP Reset
		-Analyze
		-Reboot
		-Network Diagnosis
		-Secure Network
		-Hacking
		-Encryption
		-Decryption
		-System Upgrade
		-System Update
	
	Weapons:
		-None

**Comment by jonpchin** [comment]
Here are some things we need to think about before we go actually writing any code:

1. we need to decide what programming language we are using.

2. Decide how we are going to make our map/world/rooms. Using static arrays or dynamic arrays? 2D or 3D arrays?

3. We will need to have some sort of File I/O to "save" the user's character. So all the information will be saved into a textfile periodically or when the user requests to save his game. This means at the start of the game user will need to enter "user" and "pass" and choose his class, stats, etc.

4. We need to figure out how input will be parsed whenever a user types anything into the terminal. If this is done with C or C++ there will be a lot of security issues to consider with stdin. Things like shortcuts may be handy. For example instead of user typing "east" to go east maybe allow them to type just "e".

5.We need to decide what kind of timing system we will use for this mud. This will greatly affect the combat system as well. Either tick base or balance base. Tick base will use a global game timer. So things like spells can wear off after certain amount of ticks. A tick is a unit of time in the game. Balance base is a bit more tricky but it uses the same concept of tick base but the balance will be "custom made" for each class or depending on characters afflictions or equipment their wearing. So classes which are faster will have quicker balance times.

6. Class, weapons, stat creation, equipment, monster creation, object creation and randomness. We will use an object oriented approach to handle all this. For the randomness, this could be useful for having random chance of a "rare" item dropping after a monster is killed.

7. We are going to need someone who is good with words to describe objects and how they look when people "look" at them. We could also use a story line. 

8. Quests and multiplayer!? If we ever make it this far. Maybe in the future...

**Comment by jonpchin** [comment]
Some minor things. Before anyone starts working, list what operating system you use and what coding preferences you have for example:

Operating System: Windows 8.1 and Ubuntu
Coding preferences: prefer camelCase but can also use under_scores

Prefer
int main(){

    return 0;
}
but also comfortable with:
int main()
{

    return 0;
}

**Comment by ShadyPkg** [comment]
OS: Win8.1
Coding pref: camelCase functions, under_score variables. Opening brace of function block on function line: 
int main() {
}

Busy weekend, but I'm still pondering the list of game specific questions.
I'm most comfortable with programming in C, but I'm willing to refresh on Java if we go that way. I'm not sure I can get good enough with any other language between work and school though.

**Comment by jonpchin** [comment]
We could use the help of an object oriented paradigm programming language to organize our code. Out of the 4 options I listed(C, C++, Java, and Python), C would be the worse for a game like this. 

The only way I see this working for C would be to use multiple C files to organize code and use structures as classes but there is still the issue of how we will customize stdin to parse different kinds of user input and handle incorrect user input in a simple and secure fashion. Since you mentioned Java then that is what we will have to go with. I am also fine with coding it in Python. C++ has similar issues as C but its a slightly better option then C as they have OOP features. 

So Java it is. You will have plenty of time to refresh on Java. If you want to program in C, the chatroom project that I just opened up yesterday will be written in C. And I chose C for that project only because I am curious to explore the networking and multithreading library of C. 

*Note if you can write a very sophisticated function to parse user input and to handle all kinds of special input/incorrect input in C in the new few days as well as provide documentation on how to use that function with fgets(scanf will most likely fail and getchar() may be a little complex to work with), then we can do this project in C.

**Comment by jonpchin** [comment]
The best idea I seen in C with parsing input is to use fgets and then strtol library function or sscanf to parse input. 

To give you an idea of what I am talking about:
http://faq.cprogramming.com/cgi-bin/smartfaq.cgi?id=1043284392&answer=1048382193

**Comment by jonpchin** [comment]
I been thinking about the map. Instead of using a dynamic array like you suggested I am thinking using a 2D static array for the directions like north, south, west and east. And if a player goes up or down or into a secret room then they will jump into a new 2D array. So we will have hundreds of different two arrays. This will be better then having just 1 multi dimensional dynamic array because we want O(1) access to every single room. Plus the implementation will be easier. If we ever wanted to add more rooms on the same map we can just change the 2D array dimensions or even have a 
dynamic array because we want O(1) access to every single room.  

And I've decided we are going to be using Java for this project because we need OOP features.

**Comment by ShadyPkg** [comment]
Wouldn't a linked list still be O(1), assuming the player is only ever going to visit a room linked directly to the room their in? Having O(1) static array access would work for a player that has no physical limitations of travel (teleporting around the world in the blink of an eye). 2D static arrays seems like it would be more difficult to deal with in the long term. You could just have a struct representing a room, with a room title, description, grid coordinates, id# (if the coordinates have to change for some reason),      a list of objects, a list of character entities, and a list of pointers to other valid rooms accessible from the current one. You can still get the object oriented design out of C with that. A character would be a struct too, with similar set of fields.

**Comment by jonpchin** [comment]
Teleport is one of the skills for the class Time Traveler. But you are right in that a player will have O(1) access regardless if its a link list or static array if the player was moving from one room to another room.

But also keep in mind we are not just looking at the perspective of the player. We also need to be generate/spawn random enemies/monsters and static array would be nice for that.  So if we create a link list of a map[100][100] and we wanted to put a certain boss at location map[37][79] we would have to travel through each of the nodes, and how do we know which one is the quickest route(some rooms may only have 1 or 2 exits or may be a dead end)? We would have to program extra algorithms to find the shortest path or brute force which can take a long time because multiple exits per room and some rooms have no exits at all. 

So essentially its not for the player(besides teleportation skills) but also for the us the developers, we want to access the rooms quickly. 

The struct in C idea is very similar to a class in Java. The only issue is that structs in C can't have functions. Java classes can have functions. Meaning I cannot do this in C:

struct Cybrg{
   int health = 100;
   int getHealth(int health){
             return health;
   };
}

In Java you can do what is above and call the function getHealth(). Basically it will be a lot more organized to implement all the class skills because we can store functions within classes(structs).

**Comment by ShadyPkg** [comment]
Hmm, good points. But what if multiple classes share the same skill? Wouldn't it just be better to have the skill functions be independent of the character classes, and before it's called there's a check to see if the particular player is allowed to run skill or not? Would make debugging and expanding the skills easier in the long term. For the world, perhaps we can do both? I was thinking there would be a world struct with the finite static array traveled through by x, y, z grid coordinates and each element points to a room struct. The rooms themselves could still be linked to each other with a linked list, which would allow for the existence of portals and instant room travel for monster spawn and admin functions. Best of both worlds maybe?

**Comment by jonpchin** [comment]
That is true multiple classes can share the same skill. But OOP solves that problem as well by using inheritance. For example say Cyborg and Time Traveller share the skill "Teleport" then we can make a interface or class(depending on the situation) called cyborgTimeTraveller which contains all the skills shared by Cyborg and Time Traveller. Then we can have those two classes implement(interface) or extend(class) cyborgTimeTraveller. This may even save us from having to copy and paste duplicate skills.

Another thing that may arise is that the same skill for two different classes may perform slightly different. Teleport for a Time Traveller may cost 95 energy points while Teleport for a cyborg may cost 80 energy points. So they don't exactly do the same thing.

**Comment by ShadyPkg** [comment]
Aaah, okay. Gotcha. For the teleport skill though, wouldn't it be the same code just different inputs based off the input of the player which checks their stats?

**Comment by jonpchin** [comment]
Yes teleport skill should be the same code even for different classes. However, there are hundreds of skills and there is a possibility that identical skills among two different classes are slightly different(maybe to balance a class skill if its overpowered for a certain class and not another class). 

And skills will be locked to a player's class and level. As a player levels up he gets access to more skills. Though in the past I hated the leveling system. I prefer to have players unlock all skills at level 1. But this is my preference.

**Comment by ShadyPkg** [comment]
Sorry, I keep editing my comments as you're responding. I'll try to finalize further comments before posting them. ;P

**Comment by jonpchin** [comment]
We need an actually name for this MUD. We can't just call it Sci-Fi MUD.

**Comment by jonpchin** [comment]
We either have to recruit a storywriter to our organization group or we have to write the story ourselves. I am fine with writing it but it will take a lot of my time away from coding the game. Do you know anyone who is good at story writing and likes Sci-Fi stuff like this?

**Comment by jonpchin** [comment]
Also every legit MUD has a splash screen when connecting to terminal. We don't need an artist for to make text based art but keep that in mind.  We need to customize login and store information about user, pass and his class and all his equpiment, level, etc either in a database or a textfile. Textfile will be simple to implement. We also need to really work on parsing user input. I hope you are okay with Java. If not you still have a lot of time to learn.

**Comment by ShadyPkg** [comment]
Digital Wasteland, New Eden, Outer Collective, Dustfall? 

Nah, the guys I know that would be good for the creative aspect are all currently swamped with full-time work or their own side projects. =/

**Comment by ShadyPkg** [comment]
Yea, I'm fine with Java. It's something I need to practice; been a few years since I've done anything meaningful with it. Yea, writing the user info from internal data structure to text file sounds good. Probably only necessary when turning off and on the server/game though?

**Comment by jonpchin** [comment]
Digital Wasteland sounds pretty cool. I just looked up UCF club/organization directory on knightlink. There is a Graduate Writers Organization. I am pretty sure someone in that club likes Sci-Fi.

**Comment by ShadyPkg** [comment]
Perhaps we can just use an already existing MUD client? https://en.wikipedia.org/wiki/Comparison_of_MUD_clients

I've only ever really used Pueblo, don't know much about the others.

**Comment by ShadyPkg** [comment]
Looks like TinTin++ and Potato are still active projects.

**Comment by ShadyPkg** [comment]
Btw, did you see this on the CS FB group? : http://www.neweggbusiness.com/premier-offer-student/

**Comment by jonpchin** [comment]
I have used Mudlet and Zmud. Why is the MUD client so important? Am I missing something? We have a while before we can test our game on a MUD client.

**Comment by ShadyPkg** [comment]
Thought it might save us time by getting the interface out of the way so we can focus on the game. Something we'd have to deal with eventually if we're not writing our own client, might as well design around it?

**Comment by jonpchin** [comment]
No, we won't be writing a MUD client. That is actually a GUI part. We can do that if you want but that is for later. We need to focus on back end.

**Comment by jonpchin** [comment]
I googled around. Digital Wasteland doesn't seem copyrighted. The name isn't that great but we can always change it when we think of something better.

**Comment by jonpchin** [comment]
I'll be writing the file I/O and the login to save user name and pass, class,etc during the terminal connection. If you start writing code post here so we don't both do the same thing by accident.
